### DISCUSSION TITLE
Can we operate manual control mode for specified distance

### DISCUSSION THREAD

**Discussion by Akash-Kansagara** [rand256/valetudo#487] [Q&A]
I want to use my robot to follow a specific path specified by me. Like it can be done in manual control mode. But it's like it moves a small specified distance when pressed the arrow button once. I would like to plan the route for example from one room to another and it would follow that planned route.
Any ideas if this can be done?

**Comment by rand256** [comment]
Out of boredom I've wrote [a bit of new code](https://github.com/rand256/valetudo/commit/e2ed176381a0b4561125b7cf163b089e8c95366a) recently and probably it may help.

Excerpt from MQTT kinda "manual"-like examples:
```
	/**
	 * send a RC mode command to the device
	 * starting and stopping RC mode is done automatically, you can try sending multiple commands in a row
	 * {
	 *   "command": "remote_control",
	 *   "angle": -1.3,       (between -3.14 and 3.14)
	 *   "velocity": 0.0,     (between -0.3 and 0.3)
	 *   "duration": 1000     (in ms)
	 *   "startdelay": 7500   (optional, in ms)
	 * }
	 * startdelay is the delay the device needs to wait after starting RC mode to be able to actually perform RC commands (default: 8 seconds, maybe your device would be faster)
	 * see https://github.com/marcelrv/XiaomiRobotVacuumProtocol/blob/master/rc.md for other parameters which are passed directly to `app_rc_move` miio command
	 */
```

There's also corresponding REST API that accepts alike parameters via PUT request to [/api/set_manual_control](https://github.com/rand256/valetudo/commit/e2ed176381a0b4561125b7cf163b089e8c95366a#diff-8ec1a41d135aee4de843cc696bf48cb194fe4cb0ec04f23268cb47c3f5957b50).

Both should work like this: you just put some RC commands via either way when the device is idle, and it then automatically turns on the RC mode, runs provided commands one by one, and then turns off RC mode after a small timeout when no new commands arrive. Yeah, angle/velocity/duration parameter values aren't rather very clear, but that's what manufacturer made them to be.

Oh, and here is the compiled [new test build](https://gv.erinome.net/valetudo-20220808.tar.gz) binary containing all these (and some other) changes. 

PS: It's much larger by size than my previous releases because those used my custom v12 node.js build, and this one is based on much newer but plain default v16 one that doesn't have any size optimizations. Maybe at some point I'll be able to rebuild it properly, but actually it doesn't change a lot.

**Reply by Akash-Kansagara** [reply]
Many thanks for the response. I was trying to implement the solution that you suggested, but am unable to do so.
I'm currently using the release 2022.08.0
I don't understand where should I input the commands?
Can you give me a brief idea about how and where should I type and execute the command?
Like in the REST API list it does mention "api/set_manual_control" by "PUT". But where should I write the command giving those instructions?
I am not a programmer so I don't know this stuff in detail.
And to write a input instructions like your example that you gave the picture of, will it work in the above mentioned release or it is some other version?
The main thing that I'm unable to do is: where to write the input parameters.
Thanks in Advance

**Reply by rand256** [reply]
> I'm currently using the release 2022.08.0

I'm glad for you but I have to insist that you have written your request in **valetudo RE** repository, and here I am and my releases (latest "officially" published is 0.10.7). If you're interested in non-RE versions (which, obviously, are currently became newer/better than mine in some aspects), then please go and ask anything in **their** repository.

> Can you give me a brief idea about how and where should I type and execute the command?

I honestly don't know which automation system you do use and which possibilities they provide to run commands.
Generally, there are two options:

1. MQTT. You connect your device and your automation system to some MQTT broker and send there commands, examples of which for manual control are in my previous post (and MQTT.js source file).
2. HTTP. You directly send HTTP PUT request to your device at `/api/auto_manual_control` path with a JSON payload, which should be an object like `{"angle": 0.0, "velocity": 0.1, "duration": 2000}`.

And... That's it.

> will it work in the above mentioned release or it is some other version?

As I already wrote, I've made a _special_ test build of valetudo RE with these new functions I've prepared just for you. So, no, it's not a part of any existing release. To use it I could only suggest to install the latest valetudo RE release 0.10.7 and then replace the valetudo binary with the test build I've posted above.

**Reply by Akash-Kansagara** [reply]
Oh I'm very sorry as I didn't recognise that I was asking the question in the wrong repository.
I just intend to use the software that would allow me to control the robot as I asked at the very beginning of my query. I have decided to install your mentioned RE build as it is easier to do my required tasks.
Can you provide me the steps to install your version instead of existing non-RE version.
I went through your guides, but I didn't fully understand what to do.
Thanks in Advance.
